Source/Core/DolphinWX/Debugger/CodeWindow.cpp

```diff
@@ -316,34 +314,40 @@ void CCodeWindow::StepOver()
       PowerPC::breakpoints.ClearAllTemporary();
       PowerPC::breakpoints.Add(PC + 4, true);
       CPU::EnableStepping(false);
-      JumpToAddress(PC);
-      Update();
     }
     else
     {
       SingleStep();
     }
-
-    UpdateButtonStates();
-    // Update all toolbars in the aui manager
-    Parent->UpdateGUI();
   }
 }
 
+// Returns true on a blr or on a bclr that evaluates to true.
+static bool WillInstructionReturn(UGeckoInstruction inst)
+{
+  bool counter = (inst.BO_2 >> 2) != 0 || (CTR != 0) != ((inst.BO_2 >> 1) != 0);
+  bool condition = (inst.BO_2 >> 4) != 0 || (GetCRBit(inst.BI_2) == ((inst.BO_2 >> 3) & 1));
+  bool isBclr = (inst.OPCD_7 == 0b010011) && ((inst.hex >> 1) & 0b10000) != 0;
+  return isBclr && counter && condition;
+}
+
 void CCodeWindow::StepOut()
 {
   if (CPU::IsStepping())
   {
     CPU::PauseAndLock(true, false);
     PowerPC::breakpoints.ClearAllTemporary();
 
-    // Keep stepping until the next blr or timeout after one second
+    // Keep stepping until the next return instruction or timeout after one second
     u64 timeout = SystemTimers::GetTicksPerSecond();
     u64 steps = 0;
-    PowerPC::CoreMode oldMode = PowerPC::GetMode();
+    PowerPC::CoreMode old_mode = PowerPC::GetMode();
     PowerPC::SetMode(PowerPC::MODE_INTERPRETER);
     UGeckoInstruction inst = PowerPC::HostRead_Instruction(PC);
-    while (inst.hex != 0x4e800020 && steps < timeout)  // check for blr
+    // Loop until either the current instruction is a return instruction with no Link flag
+    // or a breakpoint is detected so it can step at the breakpoint.
+    while (!(WillInstructionReturn(inst) && !inst.LK_3) && steps < timeout &&
```

Comment:
```inst.LK_3``` check should be in ```WillInstructionReturn``` function. Even if it's an instruction that will branch then return, it won't return first (i.e will branch) so the function name is perfectly accurate.